File: sns/src/main/scala/models/Endpoint.scala
package aws.sns

/**
 * An endpoint to receive notifications.
 */
sealed trait Endpoint {
  def value: String
  def protocol: String
}

object Endpoint {

  def apply(endpoint: String, protocol: String) = protocol match {
    case "http" => Http(endpoint)
    case "https" => Https(endpoint)
    case "email" => Email(endpoint)
    case "email-json" => EmailJson(endpoint)
    case "sms" => SMS(endpoint)
    case "sqs" => SQS(endpoint)
    case _ => throw new IllegalArgumentException("Invalid protocol " + protocol)
  }

  case class Http(url: String) extends Endpoint {
    override val value = url
    override val protocol = "http"
  }

  case class Https(url: String) extends Endpoint {
    override val value = url
    override val protocol = "https"
  }

  case class Email(email: String) extends Endpoint {
    override val value = email
    override val protocol = "email"
  }

  case class EmailJson(email: String) extends Endpoint {
    override val value = email
    override val protocol = "email-json"
  }

  case class SMS(phoneNumber: String) extends Endpoint {
    override val value = phoneNumber
    override val protocol = "sms"
  }

  case class SQS(queueArn: String) extends Endpoint {
    override val value = queueArn
    override val protocol = "sqs"
  }

}

File: sns/src/main/java/com/pellucid/aws/sns/Endpoint.java
package com.pellucid.aws.sns;

public class Endpoint {

    private String protocol;

    private String value;

    public Endpoint(String protocol, String value) {
        this.protocol = protocol;
        this.value = value;
    }

    public String protocol() {
        return this.protocol;
    }

    public String value() {
        return this.value;
    }

    public aws.sns.Endpoint toScala() {
        return aws.sns.Endpoint$.MODULE$.apply(protocol, value);
    }

}
